barretenberg/cpp/src/barretenberg/ecc/fields/field_impl.hpp

```diff
@@ -687,11 +688,10 @@ template <class T> field<T> field<T>::random_element(numeric::RNG* engine) noexc
     if (engine == nullptr) {
         engine = &numeric::get_randomness();
     }
-
-    uint512_t source = engine->get_random_uint512();
-    uint512_t q(modulus);
-    uint512_t reduced = source % q;
-    return field(reduced.lo);
+    constexpr field pow_2_256 = field(uint256_t(1) << 128).sqr();
```

Comment:
How sensitive are we to performance here? If we wanted to maximize performance, I think the following algorithm would be fastest:

1. generate random u512 `t`
2. feed `t` directly into a Barrett reduction algorithm, producing output `t*2^{-256} mod p`
3. multiply the output by 2^{256} mod p to return `t mod p`

The above requires 2 modular reductions. The current method performs 3, due to converting two random u256 values into `field` elements (along with additional arithmetic operations)

Very minor point, I don't know how performance sensitive we are to this algorithm

Reply:
Our old modulo algorithm was super slow, so this is a significant improvement. I tried checking what kind of an improvement would not converting to montgomery for 256-bit chunks make and it was minimal. The problem is that now that the division algorithm is not extremely slow, the bottleneck is sampling the uint512_t.

Reply:
Also we don't do 2^256 montgomery in wasm. It's different(

Reply:
thanks for the context about the wasm. Replace `2^{256}` with whatever the montgomery `R` parameter is and I think the above all applies.

But yes I appreciate it doesn't affect performance too much, it's the difference between 3 reductions and 2. It makes sense that generating the random bytes would be the limiting factor after your changes.